File: LeetCode_ProblemSet/Problem416.cpp
#include <vector>

class Solution {
   public:
    bool canPartition(std::vector<int>& nums) {
        int n = nums.size(), sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 != 0) return false;

        int target = sum / 2;
        std::vector<bool> dp(target + 1);
        dp[0] = true;
        for (int num : nums) {
            // going forward risks using same element twice
            for (int i = target; i >= num; i--) {
                if (dp[i - num]) {
                    dp[i] = true;
                }
            }
        }
        return dp[target];
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem416.rs
impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let mut sum = 0;
        for &num in &nums {
            sum += num;
        }
        if sum % 2 != 0 {
            return false;
        }

        let target = sum / 2;
        let mut dp = vec![false; (target + 1) as usize];
        dp[0] = true;
        for &num in &nums {
            for i in (num..(target + 1)).rev() {
                if dp[(i - num) as usize] {
                    dp[i as usize] = true;
                }
            }
        }
        return dp[target as usize];
    }
}
